Lib/test/test_traceback.py

```diff
@@ -457,6 +457,45 @@ def f_with_multiline():
         result_lines = self.get_exception(f_with_multiline)
         self.assertEqual(result_lines, expected_f.splitlines())
 
+    def test_caret_multiline_expression_syntax_error(self):
+        # Make sure an expression spanning multiple lines that has
+        # a syntax error is correctly marked with carets.
+        code = textwrap.dedent(r"""
+        def foo(*args, **kwargs):
+            pass
+
+        a, b, c = 1, 2, 3
+
+        foo(a, z \
+                for z in \
+                    range(10), b, c)
+        """)
+
+        def f_with_multiline():
+            # Need to defer the compilation until in self.get_exception(..)
+            return compile(code, "?", "exec")
+
+        lineno_f = f_with_multiline.__code__.co_firstlineno
+
+        expected_f = (
+            'Traceback (most recent call last):\n'
+            f'  File "{__file__}", line {self.callable_line}, in get_exception\n'
+            '    callable()\n'
+            '    ^^^^^^^^^^\n'
+            f'  File "{__file__}", line {lineno_f+2}, in f_with_multiline\n'
+            '    return compile(code, "?", "exec")\n'
+            '           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n'
+        )
+        if not isinstance(self, CPythonTracebackErrorCaretTests):
```

Comment:
Why do you need this?

Reply:
The _testcapi.traceback_print does not return as much detail, the details end with the compile() call.

Reply:
Why don't the other tests need to do this?

Reply:
The don't use compile() as they are not checking syntax errors.

Reply:
To expand on my previous answer.

Testing a syntax error is by its very nature challenging as the test cannot have the syntax error written as regular Python code. Any syntax error in the test file will stop the file being parsed correctly.

Most of the syntax tests within test_traceback.py module are within the TracebackCases class. These tests put the code with syntax errors into strings that are then compiled using the compile() function. For example:

```
    def syntax_error_with_caret_range(self):
        compile("f(x, y for y in range(30), z)", "?", "exec")
```

This is quite similar to the new test. The output is tested in this code:

```
        err = self.get_exception_format(self.syntax_error_with_caret_range,
                                        SyntaxError)
        self.assertIn("^", err[2]) # third line has caret
        self.assertEqual(err[2].count('\n'), 1)   # and no additional newline
        self.assertEqual(err[1].find("y"), err[2].find("^"))  # in the right place
        self.assertEqual(err[2].count("^"), len("y for y in range(30)"))
```

This works but is not too easy to read. Additionally only speciific parts of the traceback string are checked.

In contrast the tests within the TracebackErrorLocationCaretTests class have a nice structure where the whole output string is tested. IMO the test is also easier to read and understand. This is a typical example:
```
    def test_basic_caret(self):
        def f():
            raise ValueError("basic caret tests")

        lineno_f = f.__code__.co_firstlineno
        expected_f = (
            'Traceback (most recent call last):\n'
            f'  File "{__file__}", line {self.callable_line}, in get_exception\n'
            '    callable()\n'
            '    ^^^^^^^^^^\n'
            f'  File "{__file__}", line {lineno_f+1}, in f\n'
            '    raise ValueError("basic caret tests")\n'
            '    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n'
        )
        result_lines = self.get_exception(f)
        self.assertEqual(result_lines, expected_f.splitlines())
```
Note how the whole output string is tested.

I added the new test in the TracebackErrorLocationCaretTests class to take advantage of the nicer (IMO) test structure.

```
    def test_caret_multiline_expression_syntax_error(self):
        # Make sure an expression spanning multiple lines that has
        # a syntax error is correctly marked with carets.
        code = textwrap.dedent(r"""
        def foo(*args, **kwargs):
            pass

        a, b, c = 1, 2, 3

        foo(a, z \
                for z in \
                    range(10), b, c)
        """)

        def f_with_multiline():
            # Need to defer the compilation until in self.get_exception(..)
            return compile(code, "?", "exec")

        lineno_f = f_with_multiline.__code__.co_firstlineno

        expected_f = (
            'Traceback (most recent call last):\n'
            f'  File "{__file__}", line {self.callable_line}, in get_exception\n'
            '    callable()\n'
            '    ^^^^^^^^^^\n'
            f'  File "{__file__}", line {lineno_f+2}, in f_with_multiline\n'
            '    return compile(code, "?", "exec")\n'
            '           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n'
        )
        if not isinstance(self, CPythonTracebackErrorCaretTests):
            expected_f += (
                '  File "?", line 7\n'
                '    foo(a, z \\\n'
                '           ^^^'
                )

        result_lines = self.get_exception(f_with_multiline)
        self.assertEqual(result_lines, expected_f.splitlines())
```
The difference with the other tests within the TracebackErrorLocationCaretTests class is that test_caret_multiline_expression_syntax_error is the only test for a syntax error, therefore the compile() function is used to create a function that will raise a syntax error.

Further down in the test file the class CPythonTracebackErrorCaretTests is defined. This runs the TracebackErrorLocationCaretTests  tests again but using the traceback_print() function from _testcapi to get teh exception details. This traceback_print() function does not give any details beyond the call into the compile() function.

```
@cpython_only
@requires_debug_ranges()
class CPythonTracebackErrorCaretTests(TracebackErrorLocationCaretTests):
    """
    Same set of tests as above but with Python's internal traceback printing.
    """
    def get_exception(self, callable):
        from _testcapi import traceback_print

```

Which is why the ` if not isinstance(self, CPythonTracebackErrorCaretTests):` is present.

Reply:
The problem is that this way the test is not covering what we need it to cover for CPythonTracebackErrorCaretTests - the multiline part is the one which is excluded.

Reply:
What would you suggest?

Reply:
It's possible to get the traceback from exception_print like below, so I think the test needs to be reorganised not to use self.get_exception. It may be that the test cannot be shared between the C and Py versions (or that some common API needs to be added to both). There are some tests that run only on traceback.py, maybe this is covered there already.
```
>>> import textwrap
>>> code = textwrap.dedent(r"""
...         def foo(*args, **kwargs):
...             pass
...         a, b, c = 1, 2, 3
...         foo(a, z \
...                 for z in \
...                     range(10), b, c)
... """)
>>> 
>>> def f_with_multiline():
...     return compile(code, "?", "exec")
... 
>>> try: 
...     f_with_multiline()
... except Exception as e:
...     exc = e
... 
>>> from _testcapi import exception_print
>>> exception_print(exc)
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "<stdin>", line 2, in f_with_multiline
  File "?", line 5
    foo(a, z \
           ^^
SyntaxError: Generator expression must be parenthesized
>>> 

```

Reply:
Thank you for the advice. I have amended get_exception() to use exception_print() and the tests look good.